File: Java/String1/FirstHalf.java
/*
 * CodingBat Java Activity "firstHalf" from String-1 section.
 * codingbat.com
 */

public class FirstHalf 
{
	private static int failures = 0;
	private static final String METHOD_NAME = "firstHalf";
	
	/*
	 * INSTRUCTIONS
	 * 
	 * Given a string of even length, return the first half. So the string 
	 * "WooHoo" yields "Woo".
	 * 
	 * firstHalf("WooHoo") -> "Woo"
	 * firstHalf("HelloThere") -> "Hello"
	 * firstHalf("abcdef") -> "abc"
	 */
	public static String firstHalf(String str)
	{
		//put your code ONLY here!
	}
	
	public static void main(String[] args)
	{
		testCase("\"WooHoo\"", "\"Woo\"");
		testCase("\"HelloThere\"", "\"Hello\"");
		testCase("\"abcdef\"", "\"abc\"");
		testCase("\"ab\"", "\"a\"");
		testCase("\"\"", "\"\"");
		testCase("\"0123456789\"", "\"01234\"");
		testCase("\"kitten\"", "\"kit\"");
		
		System.out.println();
		System.out.println(failures + " test cases failed.");
	}

	private static void testCase(String a, String result)
	{
		String r;
		try
		{
			r = ""+ "\"" + firstHalf(a.substring(1,a.length()-1)) + "\"";
			if (r.equals(""+result))
			{
				System.out.print("PASS: ");
			}
			else
			{
				System.out.println("FAIL: ");
				failures++;
			}
		}
		catch (Exception e)
		{
			r = e.getMessage();
			System.out.print("FAIL: ");
			failures++;
		}
		System.out.println(METHOD_NAME+"("+a+") -> "+r + ", EXPECTED: "+result);
	}

}

File: Python/string-1/first_half.py
'''
CodingBat Python Activity "first_half" from string-1.
codingbat.com
'''

failures = 0

def first_half(str):
    '''
Given a string of even length, return the first half. So the string "WooHoo" yields "Woo".


first_half('WooHoo') → 'Woo'
first_half('HelloThere') → 'Hello'
first_half('abcdef') → 'abc'
'''
    pass #Replace pass with your code to implement this function





def catchFailures(testCase, result):
    global failures
    
    try:
        assert eval(testCase) == result, "Test Case Failed: "+testCase+ " → "+str(result)
    except Exception as e:
        failures += 1
        print(e)
            


def test():
    global failures
    
    catchFailures("first_half('WooHoo')", 'Woo')
    catchFailures("first_half('HelloThere')", 'Hello')
    catchFailures("first_half('abcdef')", 'abc')
    catchFailures("first_half('ab')", 'a')
    catchFailures("first_half('')", '')
    catchFailures("first_half('0123456789')", '01234')
    catchFailures("first_half('kitten')", 'kit')

        
    if failures > 0:
        print("\r\n "+ str(failures) + " failed test case(s)   :(")
    else:
        print("Good job -- problem solved!")


test()
